### PR TITLE
[New extractor] telegram

### LINKED ISSUE #2887: Daystar support
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

France

### Example URLs

- Single Video : https://player.daystar.tv/0MTO2ITM

### Description

Hi,
I'd love to be able to download some videos from daystar.com.
Daystar seems to be using the LightCast.com player (seen when you right click on the video).

### Verbose log

```python
yt-dlp -v https://player.daystar.tv/0MTO2ITM
[debug] Command-line config: ['-v', 'https://player.daystar.tv/0MTO2ITM']
[debug] User config "/home/sebastien/.config/yt-dlp.conf": ['--format-sort', '+proto', '--ignore-errors', '--prefer-ffmpeg', '--no-warnings', '--restrict-filenames', '--add-metadata', '--write-auto-sub', '--embed-subs', '--embed-thumbnail', '--embed-metadata', '--sub-lang', 'en,fr,es,de', '--format', 'mp4[height<=?480]/mp4/best', '--output', '%(title)s__%(format_id)s__%(id)s__%(extractor)s.%(ext)s']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, err utf-8, pref UTF-8
[debug] yt-dlp version 2022.02.04 [c1653e9ef]
[debug] Python version 3.8.10 (CPython 64bit) - Linux-5.13.0-30-generic-x86_64-with-glibc2.29
[debug] exe versions: ffmpeg 4.4.1-static (setts), ffprobe 4.4-1ppa1, phantomjs ., rtmpdump 2.4
[debug] Optional libraries: Cryptodome, secretstorage, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://player.daystar.tv/0MTO2ITM
[generic] 0MTO2ITM: Requesting header
[generic] 0MTO2ITM: Downloading webpage
[generic] 0MTO2ITM: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://player.daystar.tv/0MTO2ITM
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/yt_dlp/YoutubeDL.py", line 1381, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/yt_dlp/YoutubeDL.py", line 1451, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3/dist-packages/yt_dlp/extractor/common.py", line 612, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3/dist-packages/yt_dlp/extractor/generic.py", line 3986, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://player.daystar.tv/0MTO2ITM
```

### LINKED ISSUE #2910: Support for Telegram channels (t.me)
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

World

### Example URLs

- Single video: https://t.me/europa_press/613

### Description

Most content in Telegram channels, including videos that aren't too big, can be accessed directly via browser from URLs. So it would be interesting to be able to download them with `yt-dlp`.

It was suggested to use the "context" view, i.e. the URL in this form: https://t.me/s/europa_press/613. However, the `generic` extractor downloads a playlist of several videos, and not the specific video. Moreover, with this method, `yt-dlp` does not retrieve the metadata (there are on Telegram most of the time, including the date of creation of the video).

### Verbose log

```shell
[debug] Command-line config: ['-vU', 'https://t.me/europa_press/613']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19042-SP0
[debug] exe versions: ffmpeg 2021-09-08-git-5e7e2e5031-full_build-www.gyan.dev (setts), ffprobe 2021-09-08-git-5e7e2e5031-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [generic] Extracting URL: https://t.me/europa_press/613
[generic] 613: Requesting header
WARNING: [generic] Could not send HEAD request to https://t.me/europa_press/613: HTTP Error 501: Not Implemented
[generic] 613: Downloading webpage
WARNING: [generic] Falling back on generic information extractor.
[generic] 613: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://t.me/europa_press/613
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1381, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1451, in __extract_info
  File "yt_dlp\extractor\common.py", line 612, in extract
  File "yt_dlp\extractor\generic.py", line 3986, in _real_extract
yt_dlp.utils.UnsupportedError: Unsupported URL: https://t.me/europa_press/613
```

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Close #2910

### CODE REVIEW

**Comment by pukkandan** [comment]
Split this into 2 PRs

File: yt_dlp/extractor/telegram.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            'title': title,
+            'description': description,
+            'thumbnail': thumbnail,
+            'formats': [
+                {
+                    'ext': 'mp4',
+                    'format_id': video_id,
+                    'url': self._proto_relative_url(source),
+                    'vcodec': 'none',
+                }
```
Comment (Ashish0804) [inline yt_dlp/extractor/telegram.py:43]: Don't inline this

File: yt_dlp/extractor/telegram.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        thumbnail = self._search_regex(r'tgme_widget_message_video_thumb".*?background\-image\:url\(\'(.*?)\'\)',
+                                       webpage_embed, 'thumbnail')
+
+        source = self._search_regex('<video.*?src=\"(.*?)\"', webpage_embed, 'source')
+
+        return {
+            'id': video_id,
+            'title': title,
+            'description': description,
+            'thumbnail': thumbnail,
```
Comment (Ashish0804) [inline yt_dlp/extractor/telegram.py:36]: inline

File: yt_dlp/extractor/telegram.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+class TelegramEmbedIE(InfoExtractor):
+    IE_NAME = 'telegram:embed'
+    _VALID_URL = r'https?:\/\/t\.me\/(?P<channel_name>.*?)\/(?P<id>\d+)($|\W)'
+    _TEST = {
+        'url': 'https://t.me/europa_press/613',
+        'info_dict': {
+            'id': '613',
+            'ext': 'mp4',
+            'title': 'Europa Press',
```
Comment (Ashish0804) [inline yt_dlp/extractor/telegram.py:12]: add rest of the fields

File: yt_dlp/extractor/telegram.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        description = self._html_search_meta(['og:description', 'twitter:description'], webpage, fatal=True)
+
+        webpage_embed = self._download_webpage(f'{url}?embed=1', video_id=video_id)
+
+        thumbnail = self._search_regex(r'tgme_widget_message_video_thumb".*?background\-image\:url\(\'(.*?)\'\)',
+                                       webpage_embed, 'thumbnail')
+
+        source = self._search_regex('<video.*?src=\"(.*?)\"', webpage_embed, 'source')
+
+        return {
```
Comment (Ashish0804) [inline yt_dlp/extractor/telegram.py:32]: call sort_formats before returning

File: yt_dlp/extractor/telegram.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    _VALID_URL = r'https?:\/\/t\.me\/(?P<channel_name>.*?)\/(?P<id>\d+)($|\W)'
+    _TEST = {
+        'url': 'https://t.me/europa_press/613',
+        'info_dict': {
+            'id': '613',
+            'ext': 'mp4',
+            'title': 'Europa Press',
+            'description': 'Así de rápido fluye la lava en La Palma. Las autoridades destacan el "importante" avance '
+                           'de la colada sur (un kilómetro y medio en solo 24 horas), que va hacia la costa quemando '
+                           'invernaderos y viviendas bit.ly/3pRQv9J ',
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:15]: use md5

File: yt_dlp/extractor/telegram.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+from .common import InfoExtractor
+
+
+class TelegramEmbedIE(InfoExtractor):
+    IE_NAME = 'telegram:embed'
+    _VALID_URL = r'https?:\/\/t\.me\/(?P<channel_name>.*?)\/(?P<id>\d+)($|\W)'
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:6]:
```suggestion
    _VALID_URL = r'https?://t\.me/(?P<channel_name>[^/]+)/(?P<id>\d+)'
```

File: yt_dlp/extractor/telegram.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        thumbnail = self._search_regex(r'tgme_widget_message_video_thumb".*?background\-image\:url\(\'(.*?)\'\)',
+                                       webpage_embed, 'thumbnail')
+
+        source = self._search_regex('<video.*?src=\"(.*?)\"', webpage_embed, 'source')
+
+        formats = [{
+            'ext': 'mp4',
+            'format_id': video_id,
+            'url': self._proto_relative_url(source),
+            'vcodec': 'none'}]
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:40]:
```suggestion
        formats = [{
            'ext': 'mp4',
            'url': self._proto_relative_url(source),
            'vcodec': 'none'
        }]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
the comments on the other PR also applies here

File: yt_dlp/extractor/telegram.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            'vcodec': 'none'
+        }]
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': self._html_search_meta(['og:title', 'twitter:title'], webpage, fatal=True),
+            'description': self._html_search_meta(['og:description', 'twitter:description'], webpage, fatal=True),
+            'thumbnail': self._search_regex(r'tgme_widget_message_video_thumb"[^>]+background\-image\:url\(\'(.*?)\'\)',
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:39]: Remove unnecessary escapes
```suggestion
            'thumbnail': self._search_regex(r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
```

File: yt_dlp/extractor/telegram.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            'thumbnail': r're:^https?:\/\/cdn.*?telesco\.pe\/file\/\w+',
+        },
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id=video_id)
+
+        webpage_embed = self._download_webpage(f'{url}?embed=1', video_id=video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:23]: The URL passed can be `https://t.me/europa_press/613?embed=1`, `https://t.me/europa_press/613?xxx=yyy`, `https://t.me/europa_press/613#abcd` etc. Use urllib to correctly add the embed query

File: yt_dlp/extractor/telegram.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+        source = self._search_regex('<video[^>]+src=\"([^"]+)\"', webpage_embed, 'source')
+
+        formats = [{
+            'ext': 'mp4',
+            'url': self._proto_relative_url(source),
+            'vcodec': 'none'
+        }]
+
+        self._sort_formats(formats)
```
[the remark this answers is not in the document]
Comment (Lesmiscore) [reply yt_dlp/extractor/telegram.py]: Is `'vcodec': 'none'` is valid btw?

File: yt_dlp/extractor/telegram.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        webpage_embed = self._download_webpage(f'{url}?embed=1', video_id)
+
+        formats = [{
+            'url': self._proto_relative_url(self._search_regex(
+                '<video[^>]+src="([^"]+)"', webpage_embed, 'source')),
+            'ext': 'mp4',
+            'vcodec': 'none'
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:27]: vcoded=none implies the format is audio-only

**Pull request merged**